Implement a "Recovery Mode" for recovering crashing projects during initialization

rsubtil opened the pull request:
Closes https://github.com/godotengine/godot-proposals/issues/2628

> [!note]
> This description is updated whenever any significant changes are made, and should always reflect the current state of this PR

## Overview

This adds a new special "Recovery Mode" that, as described, disables some engine features that can make a project fail to launch during startup. When this mode is launched, the following features are completely disabled:

- Tool scripts (GDScript and C#)
- Editor plugins
- GDExtension addons
- Scene restoring

This is displayed to the user as soon as the project is opened in this mode, through both a popup, and a notification:

![image](https://github.com/user-attachments/assets/c46c2c99-566e-4f53-89b4-e59724188cc3)
![image](https://github.com/user-attachments/assets/618a2bc6-28d9-4a40-a2b0-51a202582a44)

Because some of these features may be necessary for a project to work properly, this mode may put the project in an unusable/unworkable state. Since this is a temporary mode meant to allow only for some basic troubleshooting, a few other changes were done as well to reflect this:

- All launching functionality is disabled, with the run bar modified to reflect it
![image](https://github.com/user-attachments/assets/1c19d3fb-8de6-4a52-abad-8d25df94637a)
- The Asset Library and Game View plugins are disabled, since they can't really work during this mode
![image](https://github.com/user-attachments/assets/831a9184-5052-4d4b-a4e7-6af708fa2963)
- Tool scripts are displayed differently to clarify that they're not being run during this mode
![image](https://github.com/godotengine/godot/assets/6501975/6e81ba6a-b853-483c-9d38-a854330ef97d)
- The plugin configuration panel has a warning to clarify plugins do not run during this mode
![image](https://github.com/user-attachments/assets/736cea3b-26c3-46a2-95d4-300aa8198628)

## Usage

### Automatic detection

When launching a project, Godot creates a lock file at `user://.recovery_mode_lock`. This file persists during the engine initialization, and is removed when the editor starts scanning the project files, after one second.

If any crash/hang happens during initialization, this file is not removed. So, if this file exists the next time Godot attempts to open the project, then the engine knows the project failed to open last time, thus prompting the user to open it in this mode.

### CLI

This mode can be manually triggered by opening a project with the new `--recovery-mode` flag:

```
$ godot -e -path <...> --recovery-mode
```

```
$ godot --help

Godot Engine v4.4.dev.mono.custom_build.f7a0552d5 (2024-12-06 14:59:43 UTC) - https://godotengine.org
...
Run options:
...
  --recovery-mode                   E  Start the editor in recovery mode, which disables features that can typically cause startup crashes, such as tool scripts, editor plugins, GDExtension addons, and others.
...
```

### UI

When a project is opened, the project manager searches for any existing lock file; if that's the case, it opens a popup to suggest using this mode, explaining what it does and its intended usage.

![image](https://github.com/user-attachments/assets/3c01849e-0fca-4ffb-95bc-84bc9bbdc40b)

It is also possible to use this mode manually from the new options button next to the Edit button:

![image](https://github.com/user-attachments/assets/195c1841-2e15-4636-a6cb-df50b91137aa)

## Test projects

To test this functionality, I've made a test suite of projects that crash the engine on initialization with typical user scenarios:

- **GDScript tool script:** - Project with a bad tool script, which is attached to a node on the main scene.
- **C# tool script:** - Same as above, but with a C# script, which requires rebuilding the project when the script is changed.
- **Editor plugin:** - Project with a bad editor plugin, which internally has bad GDScript code.
- **GDExtension addon:** - Project with an external error in the GDExtension addon.

All of these test cases fail to open on a normal Godot session and require manual intervention to be recovered. But under recovery mode, every test **must** open successfully.

|   | GDScript tool script | C# tool script | Editor plugin | GDExtension addon |
|---|---|---|---|---|
| Crash |[tool_gdscript_crash.zip](https://github.com/godotengine/godot/files/15503597/tool_gdscript_crash.zip) |[tool_csharp_crash.zip](https://github.com/godotengine/godot/files/15503595/tool_csharp_crash.zip) |[plugin_crash.zip](https://github.com/godotengine/godot/files/15503593/plugin_crash.zip) |[gdextension_crash.zip](https://github.com/godotengine/godot/files/15503591/gdextension_crash.zip) |
| Hang  |[tool_gdscript_hang.zip](https://github.com/godotengine/godot/files/15503599/tool_gdscript_hang.zip) |[tool_csharp_hang.zip](https://github.com/godotengine/godot/files/15503596/tool_csharp_hang.zip) |[plugin_hang.zip](https://github.com/godotengine/godot/files/15503594/plugin_hang.zip) |[gdextension_hang.zip](https://github.com/godotengine/godot/files/15503592/gdextension_hang.zip) |

- **Crash:** This project has some bad code that immediately crashes the engine when opened.
- **Hang:** This project has some bad code that loops infinitely, which hangs the engine and leaves it frozen until forcibly closed.

## Future work

This is a basic implementation so far, and the usefulness of this mode should be improved with other additional features:

- ~~An automatic mechanism to detect failing scenarios and automatically prompt users to enable this mode, as described in https://github.com/godotengine/godot-proposals/issues/2628~~ (edit: implemented in this PR too)
- Being able to open/edit scenes when some dependencies are missing, since these can originate from temporarily disabled editor/GDExtension plugins.
- Adding other problematic features for recovery mode that I didn't cover (e.g. importing assets)

rsubtil committed:
Implement a "Recovery Mode" for recovering crashing/hanging projects during initialization

rsubtil commented:
> Is the yellow border for the safe mode dialog necessary? It won't even be seen by most since windows aren't embedded by default in the editor.

Good point, I don't use the default setting of separated windows :sweat_smile:

It was more of an experiment, but I believe the yellow banner at the run bar is already noticeable enough :+1:

AThousandShips commented on editor/gui/editor_run_bar.cpp:
```diff
@@ -50,7 +51,34 @@ void EditorRunBar::_notification(int p_what) {
+
 		case NOTIFICATION_THEME_CHANGED: {
+			if (Engine::get_singleton()->is_safe_mode_hint()) {
+				main_panel->add_theme_style_override("panel", get_theme_stylebox(SNAME("LaunchPadSafeMode"), EditorStringName(EditorStyles)));
+				safe_mode_panel->add_theme_style_override("panel", get_theme_stylebox(SNAME("SafeModeButton"), EditorStringName(EditorStyles)));
```
```suggestion
				main_panel->add_theme_style_override(SceneStringName(panel), get_theme_stylebox(SNAME("LaunchPadSafeMode"), EditorStringName(EditorStyles)));
				safe_mode_panel->add_theme_style_override(SceneStringName(panel), get_theme_stylebox(SNAME("SafeModeButton"), EditorStringName(EditorStyles)));
```
New constant

rsubtil replied:
Fixed, thanks :+1: (had to create a new StringName for this one)

AThousandShips replied:
What? It's right there if you rebase your branch :)

rsubtil replied:
Oh yeah, my bad, I **really** need to make a habit of always rebasing before force-pushing :sweat_smile:

AThousandShips commented on editor/gui/editor_run_bar.cpp:
```diff
@@ -370,6 +429,35 @@ EditorRunBar::EditorRunBar() {
+		safe_mode_reload_button->set_theme_type_variation("RunBarButton");
+		safe_mode_reload_button->set_focus_mode(Control::FOCUS_NONE);
+		safe_mode_reload_button->set_tooltip_text(TTR("Disable safe mode and reload the project."));
+		safe_mode_reload_button->connect("pressed", callable_mp(this, &EditorRunBar::safe_mode_reload_project));
```
```suggestion
		safe_mode_reload_button->connect(SceneStringName(pressed), callable_mp(this, &EditorRunBar::safe_mode_reload_project));
```

rsubtil replied:
Fixed

AThousandShips commented on editor/gui/editor_run_bar.cpp:
```diff
@@ -370,6 +429,35 @@ EditorRunBar::EditorRunBar() {
+		safe_mode_button->set_focus_mode(Control::FOCUS_NONE);
+		safe_mode_button->set_text(TTR("Safe Mode"));
+		safe_mode_button->set_tooltip_text(TTR("Safe Mode is enabled. Click for more details."));
+		safe_mode_button->connect("pressed", callable_mp(this, &EditorRunBar::safe_mode_show_dialog));
```
```suggestion
		safe_mode_button->connect(SceneStringName(pressed), callable_mp(this, &EditorRunBar::safe_mode_show_dialog));
```

rsubtil replied:
Fixed

Calinou commented:
Note that we could automatically prompt the user to edit a project in safe mode after an editor (not project) crash. To do so, we need to write a file in the `.godot/` folder when the editor crashes. This can be accomplished in the crash handler when `Engine.is_editor_hint()` is `true`.

When opening any project, the editor checks for the presence of this file. If the file is present, a dialog is spawned to prompt the user to open the project in safe mode. Regardless of the user's answer, the file is removed afterwards.

When opening a project from the command line, the project opens as usual, but a warning message is printed to warn the user that they can use `--safe-mode` if desired (this is done to avoid breaking automation).

rsubtil commented:
@Calinou yes, I think this an important mechanism to complement this feature. I hadn't included it yet because, actually, I wanted to confirm this detail that you've mentioned, before proceeding:

> This can be accomplished in the crash handler when `Engine::get_singleton()->is_editor_hint()` is `true`.

Do we really want this mode to be triggered on any editor crash? I think it would generate more false positives (e.g. internal engine crashes) and generate user confusion because this mode would implicitly be triggered for all types of crashes, when it can only help in troubleshooting a subset of them.

I'm more concerned in ensuring we avoid false positives, and limiting this mode to only trigger on initialization crashes helps to reduce that. The usefulness of this mode, in my opinion, is tied to situations where a user lost the ability to edit the project at all and requires some form of external intervention to recover them.

KoBeWi commented on editor/editor_node.h:
```diff
@@ -164,6 +164,7 @@ class EditorNode : public Node {
 private:
 	friend class EditorSceneTabs;
 	friend class SurfaceUpgradeTool;
+	friend class EditorRunBar;
```
Seems unnecessary.

rsubtil replied:
This is needed on EditorRunBar to call `trigger_option`, so I can make it public instead then

KoBeWi replied:
But it's already public.

rsubtil replied:
Huuum, then there might have been some extra stuff I needed from it before but after code cleanup it isn't needed anymore, I'll recheck it.

rsubtil replied:
Fixed. It was the `MenuOption` enum that was private. So, since `trigger_option` is public, I exposed the enum as well.

KoBeWi commented on editor/gui/editor_run_bar.cpp:
```diff
@@ -50,7 +51,34 @@ void EditorRunBar::_notification(int p_what) {
+				safe_mode_button->add_theme_color_override("icon_normal_color", Color(0.3, 0.3, 0.3, 1));
+				safe_mode_button->add_theme_color_override("icon_pressed_color", Color(0.4, 0.4, 0.4, 1));
+				safe_mode_button->add_theme_color_override("icon_hover_color", Color(0.6, 0.6, 0.6, 1));
+				Color dark_color = get_theme_color("safe_mode_text_color", "Editor");
```
```suggestion
				Color dark_color = get_theme_color("safe_mode_text_color", EditorStringName(Editor));
```

rsubtil replied:
Fixed

KoBeWi commented on editor/gui/scene_tree_editor.cpp:
```diff
@@ -383,8 +383,13 @@ void SceneTreeEditor::_add_nodes(Node *p_node, TreeItem *p_parent) {
 			if (scr->is_tool()) {
-				additional_notes += "\n" + TTR("This script is currently running in the editor.");
-				button_color = get_theme_color(SNAME("accent_color"), EditorStringName(Editor));
+				if (Engine::get_singleton()->is_editor_hint()) {
```
```suggestion
				if (Engine::get_singleton()->is_safe_mode_hint()) {
```

rsubtil replied:
Fixed

KoBeWi commented on main/main.cpp:
```diff
@@ -830,15 +832,16 @@ void Main::test_cleanup() {
 
+	if (engine) {
+		memdelete(engine);
+	}
```
What's this change for?

rsubtil replied:
Huuum, I had to do this as it seemed a missed case from https://github.com/godotengine/godot/pull/92733, and without this change, a segfault would now occur with the test build, because the Engine singleton is being used during unitialization code: https://github.com/godotengine/godot/pull/92563/files#diff-dc18e6ed2e71c172f95fc57c69a508fc3a3c28bb324aa7523f10d70d71f4227bR144-R148

But now that I've checked it, the PR does this exact change already..? It was merged 2 days ago, so I might just need a rebase.

rsubtil replied:
Yup, it went away with the rebase :+1:

KoBeWi commented on main/main.cpp:
```diff
@@ -1369,6 +1372,8 @@ Error Main::setup(const char *execpath, int argc, char *argv[], bool p_second_ph
 			editor = true;
 		} else if (arg == "-p" || arg == "--project-manager") { // starts project manager
 			project_manager = true;
+		} else if (arg == "--safe-mode") { // enables safe mode
```
```suggestion
		} else if (arg == "--safe-mode") { // Enables safe mode.
```
(the other comments have incorrect style)

rsubtil replied:
Fixed

rsubtil commented:
@KoBeWi should I add the auto-detect crash functionality we discussed already in this PR, or do it in another PR after this one is merged?

I think @Calinou brings a valid point in doing it right now; if this functionality is implemented, there isn't much need to add the new "Edit in Safe Mode" button to the project manager.

I don't expect this functionality to take much effort to implement, but considering my current availability, it might still take around one week to implement.

KoBeWi commented:
Well, this is not going to be merged before 4.4, so you have some time to improve it.

Calinou reviewed (commented):
Tested locally, it works as expected.

Button styling looks a bit strange when using the Black (OLED) theme preset, due to the borders being drawn:

![image](https://github.com/godotengine/godot/assets/180032/52816048-6a32-40c1-b948-a05fb0849be8)

I'd look into hiding the borders specifically for those buttons. These borders are only drawn when **Draw Extra Borders** is enabled in the Editor Settings, which is automatically enabled by the Black (OLED) editor theme preset.

rsubtil commented:
Fixed; I've removed the safe mode button's borders since it's implicitly styled already :+1:

![image](https://github.com/godotengine/godot/assets/6501975/71804655-427a-4bef-a29c-c6c7b9d87b94)

bruvzg reviewed (commented):
Looks good to me.

The `NodeWarning` icon on the "edit" button in the project manager looks an out of place. The yellow badge in the editor is reasonable, but in the PM it probably should use a monochrome icon.

rsubtil commented:
@bruvzg I agree with you, but there is actually no need for this button anymore once the crash detection mechanism is implemented, since it should be able to reliably detect and prompt for this mode only when it's relevant, which also makes for a better UX.

bruvzg commented:
Only doing it after the crash won't cover all uses cases, broken editor addon can mess up the editor to unusable state but not crash and allow closing it normally.

While I agree that having a lock file and auto asking for safe mode after crash is good (how would it tell apart multiple instances from the crash?). But keeping it in the UI is useful as well, maybe we should have a single custom button with two pressable areas, for normal edit and "down arrow" part for context menu with extra safe mode option.

rsubtil commented:
> Only doing it after the crash won't cover all uses cases, broken editor addon can mess up the editor to unusable state but not crash and allow closing it normally.

There are plenty of ways for the engine to crash due to bad user code. But this feature is not meant as an effective tool to help in all cases. With plugins, for instance, a plugin could crash only when some specific behavior is triggered, or only after 3~4 seconds, or randomly due to a data race with multi-threaded work, etc...

This feature is focused on the scenario where users lost the ability to open a project at all, and it requires some form of external intervention to recover the project. So, more specifically, for any crashes/hangs that occur during the engine initialization, before any interaction is possible.

> (how would it tell apart multiple instances from the crash?)

It currently cannot tell the difference. The current implementation I'm doing only detects the presence of the lock file to prompt the safe mode. So, in practice, it prompts it after only one initialization crash occurred.

> maybe we should have a single custom button with two pressable areas, for normal edit and "down arrow" part for context menu with extra safe mode option.

I like this idea a lot! I made a mockup of it, something like this?
![image](https://github.com/godotengine/godot/assets/6501975/13f97ec2-a33c-4a2b-bc1f-05d918d71a8d)

AThousandShips commented on editor/gui/editor_run_bar.cpp:
```diff
@@ -50,7 +51,35 @@ void EditorRunBar::_notification(int p_what) {
+				safe_mode_button->add_theme_color_override("icon_pressed_color", Color(0.4, 0.4, 0.4, 1));
+				safe_mode_button->add_theme_color_override("icon_hover_color", Color(0.6, 0.6, 0.6, 1));
+				Color dark_color = get_theme_color("safe_mode_text_color", EditorStringName(Editor));
+				safe_mode_button->add_theme_color_override("font_color", dark_color);
```
```suggestion
				safe_mode_button->add_theme_color_override(SceneStringName(font_color), dark_color);
```
New constant

rsubtil replied:
Fixed :+1:

rsubtil commented:
I've finish up the auto-detection mechanism and rework to the project manager interface. Will update the original PR description with these new changes.

m4gr3d commented on editor/project_manager.cpp:
```diff
@@ -711,14 +739,21 @@ void ProjectManager::_on_projects_updated() {
+void ProjectManager::_on_open_options_selected(int p_option) {
+	switch (p_option) {
+		case 0: // Edit in safe mode
+			_open_safe_mode_ask();
```
Missing break statement.

rsubtil commented:
Fixed a merge conflict that appeared in the meantime. But, since `master` updated to 4.4, this revealed a bug with the auto-detection mechanism: if a warning existed when opening a project (e.g. "Do you wanna upgrade the project to Godot 4.4"), this would bypass the check for safe mode.

I've changed the code to divide this in two steps: the `check_safe_mode` exclusively to check the safe mode, and the `check_warnings` for the current warnings. And then modified the code to call the correct instances depending on the current scenario. Now, if a project has both warning types, it first displays the safe mode popup, then the warnings popup before launching.

fire commented:
Remember headless mode has no prompts and needs to work with integration and deployment

rsubtil commented:
> Remember headless mode has no prompts and needs to work with integration and deployment

The detection mechanism is implemented only on the project manager; to trigger this mode from CLI, one has to pass the `--safe-mode` flag explicitly.

akien-mga commented:
Just a quick update on the review status: I'd like to do a proper review myself, that's a feature I've wanted for a while so I want to make sure it works well.

Will also need a rebase before merge.

rsubtil commented:
Thanks, I didn't notice another merge conflict arose in the meantime; I'll fix it, possibly today

rsubtil commented:
Merge conflict has now been fixed, but it wasn't as simple as I hoped; it came from https://github.com/godotengine/godot/pull/96861, @KoBeWi please have a look to ensure I didn't break anything :sweat_smile:

KoBeWi commented:
You renamed `_open_selected_projects_with_migration` method

KoBeWi commented on editor/project_manager.h:
```diff
@@ -225,6 +237,7 @@ class ProjectManager : public Control {
 	Button *full_convert_button = nullptr;
 
 	String version_convert_feature;
+	bool open_in_safe_mode;
```
```suggestion
	bool open_in_safe_mode = false;
```

rsubtil replied:
Fixed :+1:

KoBeWi commented:
> Since you're doing something similar, I saw fit to also rename it to _open_selected_project_check_migration.

But the function performs the migration. The checks are technically done before the info dialog appears.

rsubtil commented:
> But the function performs the migration. The checks are technically done before the info dialog appears.

Ok yeah, you're right, I've reverted this name change.

AThousandShips reviewed (approved):
LGTM

akien-mga commented:
I rebased to solve merge conflicts. I'll try to find a bit of time to finally test the feature and approve.
Also CC @bruvzg if you want to do another review pass.

akien-mga commented:
Tested briefly with the MRP of #99001 which causes a crash on start from a tool script, and with this PR it's properly flagged as requiring Safe Mode, which works well to open the project and then manually fix the tool script.

One minor nitpick, when you manually select Edit in Safe Mode from the Project Manager for any project, it shows the same dialog as if that project did crash previously, so it starts with:
![image](https://github.com/user-attachments/assets/b16a61dc-0a06-4d61-976a-7c7afab6a81e)

This isn't accurate if selecting Safe Mode manually on projects that didn't crash.

rsubtil commented on editor/project_manager/project_list.cpp:
```diff
@@ -486,7 +488,13 @@ ProjectList::Item ProjectList::load_project_data(const String &p_path, bool p_fa
-	return Item(project_name, description, project_version, tags, p_path, icon, main_scene, unsupported_features, last_edited, p_favorite, grayed, missing, config_version);
+	// We can't use globalize_path because there is no loaded project. So we replicate the behavior of accessing user://
+	if (!cf_project_name.is_empty()) {
+		String safe_mode_lock_file = OS::get_singleton()->get_user_data_dir(cf_project_name).path_join(".safe_mode_lock");
```
> Shouldn't it do get_safe_dir_name on a name, like it's done in the actualt project,

Yup, I missed that. I also don't remember why I used the raw `cf_project_name` variable instead of the unescaped `project_name`, so I'll revisit this.

> also seems like it won't handle application/config/use_custom_user_dir/application/config/custom_user_dir_name override

This is indeed a problem, because this query is inside the `get_user_dir`. I already had to move the query for `application/config/name` outside to be able to set it from the context of the project manager. My best idea so far is to extract all project specific logic from this function, and move this logic to the generic `os.cpp::get_user_data_dir()` function, with the OS-specific overrides now becoming quite simple:

```cpp
// os.cpp
String OS::get_user_data_dir() const {
	String appname = get_safe_dir_name(GLOBAL_GET("application/config/name"));
	if (!appname.is_empty()) {
		bool use_custom_dir = GLOBAL_GET("application/config/use_custom_user_dir");
		if (use_custom_dir) {
			String custom_dir = get_safe_dir_name(GLOBAL_GET("application/config/custom_user_dir_name"), true);
			if (custom_dir.is_empty()) {
				custom_dir = appname;
			}
			return get_user_data_dir(custom_dir);
		} else {
			return get_user_data_dir(get_godot_dir_name().path_join("app_userdata").path_join(appname));
		}
	}

	return get_user_data_dir(get_godot_dir_name().path_join("app_userdata").path_join("[unnamed project]"));
}
...

// os_unix.cpp
String OS_Unix::get_user_data_dir(const String &p_dirname) const {
	return get_data_path().path_join(p_dirname);
}
...

// os_windows.cpp
String OS_Unix::get_user_data_dir(const String &p_dirname) const {
	return get_data_path().path_join(p_dirname).replace("\\", "/");
}
``` 
What do you think?

rsubtil replied:
I ended up extracting this logic to `os.cpp`, which has the benefit of unifying 3 implementations (Unix, Windows and Web).

rsubtil commented:
> One minor nitpick, when you manually select Edit in Safe Mode from the Project Manager for any project, it shows the same dialog as if that project did crash previously, so it starts with:

Good catch! I'll hide that part of the message when this is triggered manually then :+1:

rsubtil commented:
So, in order to remove the first part of the text when the dialog is manually triggered, I created two separate nodes to simply hide the first one depending on that.

And, since now I'm using nodes instead of popup's `set_text`, I experimented with using a `RichTextLabel` for the details part, in order to properly render the bullet points:
![image](https://github.com/user-attachments/assets/36309a70-0f55-4862-8d66-e47bf47f60c8)

What do you think?

Faless commented:
Sorry for being very late to the party, I think the term `safe mode` is very misleading, and gives a false sense of security.

I suggest to rename the option as `recovery mode` since the goal is to make "crashing projects" recoverable, and not to create a secure environment.

rsubtil commented:
> I suggest to rename the option as recovery mode since the goal is to make "crashing projects" recoverable, and not to create a secure environment.

Huum I agree, it does a better job at conveying that this mode is mostly intended as a "rescue" mode, not as a slightly less functional but still usable mode for development. I'll change the naming and fix the merge conflict that appeared in the meantime :+1:

rsubtil commented:
I've made some significant changes this time, so here's a summary (the original description will be updated shortly too):
- Renamed the feature from "safe mode" to "recovery mode" as suggested on https://github.com/godotengine/godot/pull/92563#issuecomment-2508968464
- Changed mode description text to use proper bullet points (•) similarly to other cases in the project already
- Extracted all logic for determining the user directory to a generic implementation on `os.cpp` (due to https://github.com/godotengine/godot/pull/92563#discussion_r1839976559).
  - Affected OSes (Windows, Linux, Web) were adapted, while others are not affected since they implement custom behavior (macOS, Android, iOS)

And some minor changes I took the liberty of doing as well:

- Disabled the newly added "Game" tab as well during this mode, since it is not possible to run the project.
- Changed the error codes for GDExtension and EditorNode from OK to invalid ones.

Calinou reviewed (commented):
Tested locally (rebased on top of `master` https://github.com/godotengine/godot/commit/2582793d408ade0b6ed42f913ae33e7da5fb9184), it works as expected.

I noticed one issue though. If you use `--recovery-mode` when running a project (as opposed to opening it in the editor), the project will run but without any scripts running. For example, the 3D platformer demo will appear to be stuck (but with animations playing):

![image](https://github.com/user-attachments/assets/0c695a2f-a60c-4fb2-b59d-448c507898b0)

The project manager also runs with `--recovery-mode`, but with the asset library tab disabled (and not hidden like it is in the editor):

![image](https://github.com/user-attachments/assets/0d568cf4-8683-435f-a38a-54d2e5f17727)

I suggest ignoring `--recovery-mode` if the editor isn't running (i.e. ignore it in the game and project manager).

bruvzg commented on editor/gui/editor_run_bar.cpp:
```diff
@@ -425,6 +486,42 @@ EditorRunBar::EditorRunBar() {
 
+	if (Engine::get_singleton()->is_recovery_mode_hint()) {
+		recovery_mode_popup = memnew(AcceptDialog);
+		recovery_mode_popup->set_min_size(Size2(550, 70));
```
```suggestion
		recovery_mode_popup->set_min_size(Size2(550, 70) * EDSCALE); //needs #include "editor/themes/editor_scale.h" added
```
Otherwise, it will have strange size on hiDPI displays:
<img width="286" alt="Screenshot 2025-01-02 at 20 25 11" src="https://github.com/user-attachments/assets/60b0d508-83d7-40e7-8b29-c057edb7b80e" />

rsubtil replied:
Fixed, thanks :+1:

bruvzg commented on editor/project_manager.cpp:
```diff
@@ -1498,6 +1593,14 @@ ProjectManager::ProjectManager() {
 		add_child(multi_run_ask);
 
+		open_recovery_mode_ask = memnew(ConfirmationDialog);
+		open_recovery_mode_ask->set_min_size(Size2(550, 70));
```
```suggestion
		open_recovery_mode_ask->set_min_size(Size2(550, 70) * EDSCALE);
```

rsubtil replied:
Fixed

bruvzg reviewed (approved):
Quickly tested it, seems to be working as expected.

rsubtil commented:
> I suggest ignoring --recovery-mode if the editor isn't running (i.e. ignore it in the game and project manager).

Thanks for testing @Calinou, I didn't remember to check for this flag to only be applied when the editor is run. I've added a check as suggested to prevent this.

akien-mga commented:
Thanks! Awesome work 👏
